src/beast/beast/wsproto/socket.h

```diff
@@ -0,0 +1,1062 @@
+/*
+    This file is part of Beast: https://github.com/vinniefalco/Beast
+    Copyright 2013, Vinnie Falco <vinnie.falco@gmail.com>
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef BEAST_WSPROTO_SOCKET_H_INCLUDED
+#define BEAST_WSPROTO_SOCKET_H_INCLUDED
+
+#include <beast/wsproto/option.h>
+#include <beast/wsproto/detail/socket_base.h>
+#include <beast/asio/streambuf_readstream.h>
+#include <beast/http/message.h>
+#include <beast/http/string_body.h>
+#include <boost/asio.hpp>
+#if DOXYGEN
+#include <beast/wsproto/README.md>
+#endif
+
+namespace beast {
+namespace wsproto {
+
+/** WebSocket message metadata.
+*/
+struct msg_info
+{
+    /// Indicates the type of message (binary or text).
+    opcode::value op;
+
+    /// `true` if all octets for the current message are received.
+    bool fin;
+};
+
+//--------------------------------------------------------------------
+
+/** Provides message-oriented functionality using WebSocket.
+
+    The socket class template provides asynchronous and blocking
+    message-oriented functionality necessary for clients and servers
+    to utilize the WebSocket protocol.
+
+    Thread Safety
+    
+    * Distinct objects: Safe.
+
+    * Shared objects: Unsafe. The application must
+      also ensure that all asynchronous operations are
+      performed within the same implicit or explicit strand.
+
+    Example:
+
+        To use the WebSocket socket template with an
+        ip::tcp::socket, you would write:
+
+        @code
+        wsproto::socket<ip::tcp::socket> ws(io_service);
+        @endcode
+        Alternatively, you can write:
+        @code
+        ip::tcp::socket sock(io_service);
+        wsproto::socket<ip::tcp::socket&> ws(sock);
+        @endcode
+
+    @note A socket object must not be destroyed while there are
+    pending asynchronous operations associated with it.
+
+    @see AsyncReadStream, AsyncWriteStream,
+    Decorator, Streambuf, SyncReadStream, SyncWriteStream.
+*/
+template<class Stream>
+class socket : public detail::socket_base
+{
+    friend class ws_test;
+
+    Stream next_layer_;
+    streambuf_readstream<
+        std::remove_reference_t<Stream>&,
```

Comment:
Remove and then re-add the ref? (If Stream is already a ref the extra & is collapsed).